pkg/webhook/external/list.go

```diff
@@ -0,0 +1,218 @@
+package external
+
+import (
+	"context"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/pkg/expressions"
+)
+
+func (g *genericWebhookReceiver) listTargetObjects(
+	ctx context.Context,
+	target kargoapi.GenericWebhookTarget,
+	actionEnv map[string]any,
+) ([]client.Object, error) {
+	listOpts, err := buildListOptionsForTarget(g.project, target, actionEnv)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build list options: %w", err)
+	}
+	switch target.Kind {
+	case kargoapi.GenericWebhookTargetKindWarehouse:
+		warehouses := new(kargoapi.WarehouseList)
+		if err := g.client.List(ctx, warehouses, listOpts...); err != nil {
+			return nil, fmt.Errorf("error listing %s targets: %w", target.Kind, err)
+		}
+		if target.Name == "" {
+			return itemsToObjects(warehouses.Items), nil
+		}
+		name, err := evalAsString(target.Name, actionEnv)
+		if err != nil {
+			return nil, fmt.Errorf("failed to evaluate target name as string: %w", err)
+		}
+		for _, wh := range warehouses.Items {
+			if wh.Name == name {
+				return []client.Object{&wh}, nil
+			}
+		}
+		return nil, nil
+	default:
+		return nil, fmt.Errorf("unsupported target kind: %q", target.Kind)
+	}
+}
+
+// buildListOptionsForTarget builds a list of client.ListOption based on the
+// provided GenericWebhookTarget's selectors. The returned ListOptions can be
+// used to list Kubernetes resources that match the target's selection criteria.
+func buildListOptionsForTarget(
+	project string,
+	t kargoapi.GenericWebhookTarget,
+	env map[string]any,
+) ([]client.ListOption, error) {
+	listOpts := []client.ListOption{client.InNamespace(project)}
```

Comment:
A receiver defined in `ClusterConfig` should select resources from _all_ namespaces, so shouldn't this particular list option be added conditionally?